main/ZactuatorFASTLED.ino

```diff
@@ -0,0 +1,293 @@
+/*
+  OpenMQTTGateway - ESP8266 or Arduino program for home automation
+
+  This gateway enables to use RGBLED strips like WS2812
+
+    Copyright: (c)
+
+    This file is part of OpenMQTTGateway.
+
+    OpenMQTTGateway is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    OpenMQTTGateway is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "User_config.h"
+
+#ifdef ZactuatorFASTLED
+
+#ifdef ESP8266
+#include <FastLED.h>
+#else
+#include <FastLED.h>
+#endif
+#include <ArduinoJson.h>
+enum LEDState
+{
+  OFF,
+  FIRE,
+  GENERAL
+};
+LEDState currentLEDState;
+long lastUpdate = 0;
+long currentUpdate = 0;
+CRGB leds[FASTLED_NUM_LEDS];
+CRGB ledColorBlink[FASTLED_NUM_LEDS];
+StaticJsonBuffer<200> jsonBuffer;
+bool blinkLED[FASTLED_NUM_LEDS];
+const long blinkInterval = 300;
+const long fireUpdate = 10;
+unsigned long previousMillis = 0;
+CRGBPalette16 gPal;
+
+void setupFASTLED()
+{
+  trc(F("FASTLED_DATA_PIN "));
+  trc(String(FASTLED_DATA_PIN));
+  trc(F("FASTLED_NUM_LEDS "));
+  trc(String(FASTLED_NUM_LEDS));
+  trc(F("ZactuatorFASTLED setup done "));
+  FastLED.addLeds<FASTLED_TYPE, FASTLED_DATA_PIN>(leds, FASTLED_NUM_LEDS);
+}
+
+//returns the current step of the animation
+int animation_step(int duration, int steps)
+{
+  int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
+  return currentStep;
+}
+
+//returns the number of steps since the last update
+int animation_step_count(int duration, int steps)
+{
+  long lastAnimationNumber = lastUpdate / duration;
+  long currentAnimationNumber = currentUpdate / duration;
+  int lastStep = ((lastUpdate % duration) / ((float)duration)) * steps;
+  int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
+
+  return currentStep - lastStep + (currentAnimationNumber - lastAnimationNumber) * steps;
+}
+
+void FASTLEDLoop()
+{
+
+  lastUpdate = currentUpdate;
+  currentUpdate = millis();
+
+  if (currentLEDState == GENERAL)
+  {
+
+    for (int i = 0; i < FASTLED_NUM_LEDS; i++)
+    {
+
+      int count = animation_step_count(blinkInterval, 2);
+      if (count > 0)
+      {
+        int step = animation_step(blinkInterval, 2);
+
```

Comment:
place the for loop here, so that the animation functions are outside of the for loop

Reply:
> 
> 
> how to change the animation type with the Json request?
> should the animations be extensible, so you can easily add you own animations without changing the whole code?
At the moment I have no deep knowledge about animation. I only copied the code. From my point of view should this be a point for a next release.

Reply:
I update the code